Docs content for Go drivers and ORMs (#3)
nchandrappa/yugabyte-db: docs/content/latest/drivers-orms/go/pgx.md

---
title: Go Drivers
linkTitle: Go Drivers
description: Go Drivers for YSQL
headcontent: Go Drivers for YSQL
image: /images/section_icons/sample-data/s_s1-sampledata-3x.png
menu:
  latest:
    name: Go Drivers
    identifier: pgx-driver
    parent: go-drivers
    weight: 400
isTocNested: true
showAsideToc: true
---

> For Go Applications, most drivers provide database connectivity through the standard `database/sql`

Reviewer: ```suggestion

For Go Applications, most drivers provide database connectivity through the standard `database/sql`
```

> API. YugabyteDB supports [PGX Driver](https://github.com/jackc/pgx) and the

Reviewer: ```suggestion
API. YugabyteDB supports the [PGX Driver](https://github.com/jackc/pgx) and the
```

[PQ Driver](https://github.com/lib/pq).

<ul class="nav nav-tabs-alt nav-tabs-yb">

  <li >
    <a href="/latest/drivers-orms/go/pgx/" class="nav-link active">
      <i class="icon-postgres" aria-hidden="true"></i>
      PGX Driver
    </a>
  </li>

  <li >
    <a href="/latest/drivers-orms/go/pq/" class="nav-link">
      <i class="icon-postgres" aria-hidden="true"></i>
      PQ Driver
    </a>
  </li>

</ul>

The [PGX driver](https://github.com/jackc/pgx/) is one of the most popular and actively maintained
drivers for PostgreSQL which can be used for connecting to YugabyteDB YSQL as well.<br/>
This driver allows Go programmers to connect to YugabyteDB database to execute DMLs and DDLs using
the PGX APIs. It also supports the standard `database/sql` package.

## Quick Start

Learn how to establish a connection to YugabyteDB database and begin simple CRUD operations using
the steps in the [Build an application](/latest/quick-start/build-apps/go/ysql-pgx) page under the
Quick start section.

Let us break down the quick start example and understand how to perform the common tasks required
for Go App development using the PGX driver.

## Step 1: Import the Driver Package

> You can import the PGX driver package by adding the following import statement in your Go code.

Reviewer: Most places, you can omit "you can".
```suggestion
Import the PGX driver package by adding the following import statement in your Go code.
```

### Import Statement

```golang
import (
  "github.com/jackc/pgx/v4"
)
```

## Step 2: Connect to YugabyteDB

Go Apps can connect to the YugabyteDB database using the `pgx.Connect()` function.
All the common functions or structs required for working with YugabyteDB database is part of `pgx` package.

Use the `pgx.Connect()` method for getting connection object for the YugabyteDB database which can
be used for performing DDLs and DMLs against the database.

> PGX Connection url is in the format given below:

Reviewer: ```suggestion
The PGX connection URL is in the following format:
```

```golang
postgresql://username:password@hostname:port/database
```

> Code snippet for connecting to YugabyteDB:

Reviewer: For all of the connection snippets... We suggest at the top that you create a Cloud cluster to do this stuff, but then we don't specify SSL parameters. That seems like something we ought to fix. (You're going to need cert paths for all of these, too.)

Author: Yes, it is explained in details in pgx reference page, linked at the bottom.

Author: Yes, it is explained in details in pgx reference page, linked at the bottom.

                        " VALUES (1, 'John', 35, 'Go')";
_, err = conn.Exec(context.Background(), insertStmt)
if err != nil {
  fmt.Fprintf(os.Stderr, "Exec for create table failed: %v\n", err)
}
```

The pgx driver automatically prepares and caches statements by default, so that the developer does
not have to. 

### Query Data

In order to query data from YugabyteDB tables, execute the `SELECT` statement using the function
`conn.Query()`.
Query results are returned in `pgx.Rows` which can be iterated using `pgx.Rows.next()` method.
Then read the data using `pgx.rows.Scan()`.

The SELECT DML statement:

```sql
SELECT * from employee;
```

Code snippet:

```golang
var name string
var age int
var language string

rows, err := conn.Query(context.Background(), "SELECT name, age, language FROM employee WHERE id = 1")
if err != nil {
  log.Fatal(err)
}
defer rows.Close()

fmt.Printf("Query for id=1 returned: ");
for rows.Next() {
  err := rows.Scan(&name, &age, &language)
  if err != nil {
    log.Fatal(err)
  }
  fmt.Printf("Row[%s, %d, %s]\n", name, age, language)
}

err = rows.Err()
if err != nil {
  log.Fatal(err)
}
```

> See this page - [PGX Driver](/latest/reference/drivers/go/pgx-reference/) - for more details.

Reviewer: ```suggestion
See the [PGX Driver](/latest/reference/drivers/go/pgx-reference/) page for more details.
```
